lib/util.c

```diff
@@ -480,6 +481,38 @@ EXPORTED int create_tempfile(const char *path)
     return fd;
 }
 
+EXPORTED char *create_tempdir(const char *path, const char *subname)
+{
+    struct buf buf = BUF_INITIALIZER;
+    char *dbpath = NULL;
+
+    buf_setcstr(&buf, path);
+    if (!buf.len || buf.s[buf.len-1] != '/') {
+        buf_putc(&buf, '/');
+    }
+    buf_appendcstr(&buf, "cyrus-");
+    buf_appendcstr(&buf, subname && *subname ? subname : "tmpdir");
+    buf_appendcstr(&buf, "-XXXXXX");
+    buf_cstring(&buf);
+    dbpath = xstrdupnull(mkdtemp(buf.s));
+
+    buf_free(&buf);
+    return dbpath;
+}
+
+static int removedir_cb(const char *fpath,
+                        const struct stat *sb __attribute__((unused)),
+                        int typeflag __attribute__((unused)),
+                        struct FTW *ftwbuf __attribute__((unused)))
+{
+    return remove(fpath);
+}
+
+EXPORTED int removedir(const char *dbpath)
+{
+    return nftw(dbpath, removedir_cb, 128, FTW_DEPTH|FTW_PHYS);
```

Comment:
this could probably be used in a few more places, in particular some of the other xapian stuff is shelling out to `rm` I think!